openedx/features/course_experience/views/course_home_messages.py

```diff
@@ -137,11 +132,8 @@ def _register_course_home_messages(request, course_id, user_access, course_start
     # Only show the set course goal message for enrolled, unverified
     # users that have not yet set a goal in a course that allows for
     # verified statuses.
-    has_verified_mode = CourseMode.has_verified_mode(CourseMode.modes_for_course_dict(unicode(course.id)))
-    is_already_verified = CourseEnrollment.is_enrolled_as_verified(request.user, course_key)
-    user_goal = get_course_goal(auth.get_user(request), course_key) if not request.user.is_anonymous() else None
-    if user_access['is_enrolled'] and has_verified_mode and not is_already_verified and not user_goal \
-            and ENABLE_COURSE_GOALS.is_enabled(course_key) and settings.FEATURES.get('ENABLE_COURSE_GOALS'):
+    user_goal = get_course_goal(auth.get_user(request), course_key)
+    if has_course_goal_permission(request, course_id, user_access) and not user_goal:
```

Comment:
Nice clean-up.  As mentioned in the other comment, you might need to break out 1 or 2 of these cases.